pkg/apis/networking/v1alpha1/clusteringress_types.go

```diff
@@ -0,0 +1,379 @@
+// - Timeout & Retry can be configured.
+// - Headers can be appended.
+type IngressSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// TLS configuration. Currently the ClusterIngress only supports a single TLS
+	// port, 443. If multiple members of this list specify different hosts, they
+	// will be multiplexed on the same port according to the hostname specified
+	// through the SNI TLS extension, if the ingress controller fulfilling the
+	// ingress supports SNI.
+	// +optional
+	TLS []ClusterIngressTLS `json:"tls,omitempty"`
+
+	// A list of host rules used to configure the ClusterIngress.
+	// +optional
+	Rules []ClusterIngressRule `json:"rules,omitempty"`
+}
+
+// ClusterIngressTLS describes the transport layer security associated with an ClusterIngress.
+type ClusterIngressTLS struct {
+	// Hosts are a list of hosts included in the TLS certificate. The values in
+	// this list must match the name/s used in the tlsSecret. Defaults to the
+	// wildcard host setting for the loadbalancer controller fulfilling this
+	// ClusterIngress, if left unspecified.
+	// +optional
+	Hosts []string `json:"hosts,omitempty"`
+
+	// SecretName is the name of the secret used to terminate SSL traffic.
+	SecretName string `json:"secretName,omitempty"`
+
+	// SecretNamespace is the namespace of the secret used to terminate SSL traffic.
+	SecretNamespace string `json:"secretNamespace,omitempty"`
+
+	// ServerCertificate identifies the certificate filename in the secret.
+	// Defaults to `tls.cert`.
+	// +optional
+	ServerCertificate string `json:"serverCertificate,omitempty"`
+
+	// PrivateKey identifies the private key filename in the secret.
+	// Defaults to `tls.key`.
+	// +optional
+	PrivateKey string `json:"privateKey,omitempty"`
+}
+
+// ClusterIngressRule represents the rules mapping the paths under a specified host to
+// the related backend services. Incoming requests are first evaluated for a host
+// match, then routed to the backend associated with the matching ClusterIngressRuleValue.
+type ClusterIngressRule struct {
+	// Host is the fully qualified domain name of a network host, as defined
+	// by RFC 3986. Note the following deviations from the "host" part of the
+	// URI as defined in the RFC:
+	// 1. IPs are not allowed. Currently a ClusterIngressRuleV can only apply to the
+	//	  IP in the Spec of the parent ClusterIngress.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
+	//	  Currently the port of an ClusterIngress is implicitly :80 for http and
+	//	  :443 for https.
+	// Both these may change in the future.
+	// Incoming requests are matched against the host before the ClusterIngressRuleValue.
```

Comment:
It looks like `ClusterIngressRuleValue` doesn't support matching on Host, so I'm not sure this matters?